Fix FileSystem dock won't show any file folders (reverted)

Hilderin opened the pull request:
Fixes #92335, Fixes #47053 and Fixes #85082

This PR improves:
- FileSystem Dock not showing progress bar on first file scanning
- Better progression (%) for scanning large projects with a lot of folders and files
- Performance optimization on scanning project files for large projects
- Add progression popup for _update_scan_actions step (instead of Editor freeze)
- Add progression popup for _update_script_classes step (instead of Editor freeze)
- Add progression popup for loading editor layout and reopening scenes (instead of Editor freeze)

I tested these modifications with a project having all the files from https://github.com/PMDCollab/SpriteCollab/tree/master/sprite (more than 125000 files and 385Mb).

Before:
FileSystem was empty without any feedback for a couple of minutes and after that the editor was just frozen:

![image](https://github.com/godotengine/godot/assets/81109165/75c7a53c-cf89-4df5-ad09-c94382f423a9)


After modifications:
![image](https://github.com/godotengine/godot/assets/81109165/58b70ca3-0b2a-4ea5-86a2-346b29bc1b60)
![image](https://github.com/godotengine/godot/assets/81109165/2d4c76ae-eccf-43ac-8b69-d3ac4b8ad633)


**Performance optimization on project scanning before asset importation**
For a 30000 files project, it took around 2m50s on startup to scan the project.
After a couple of optimizations, it now takes 1m30s for the same project.

Before:
![image](https://github.com/godotengine/godot/assets/81109165/4442d2d2-0fc7-4c46-b42f-51648e9e40e3)

After:
![image](https://github.com/godotengine/godot/assets/81109165/e07dbc15-b59a-4f04-9c8d-e249f9786f53)


Causes:
- The FileAccess:exists was calling the open method which has a couple of overhead over file_exists.
- The _scan_new_dir was double sorting the folders.
- The reimport_files was recalculating the import_groupe_file even if _update_scan_actions did that just before.

** Edit: Add progression popup for loading editor layout and reopening scenes.

Hilderin committed:
Fix FileSystem dock won't show any file folders

Hilderin commented:
I need your advise... if you look at the issue #85082, one of the problem was that an empty scene was present during the loading.
With this PR, this empty scene is present but a progress bar is displayed in the FileSystem Dock. My concern is that the user can modify this empty scene and do a lot of things in the editor while the project file scan is in progress.

I can let this that way, at lease now, the user as a feedback from the editor.

An easy fix for that would be to run the first scan on the matin thread, blocking the user from interacting with the editor. Anyway, if there are files to reimport, the editor does it on the main thread. I could display the progression for the scan in a popup instead of the FileSystem, but the empty scene would still be visible.

I guess the perfect solution would be to not show the editor while loading, only progress popups, but I think was would need a lot more work?

KoBeWi commented:
I tested it with my project. I got FileSystemScan, then some new Scan Actions popup, then scene group scan popup and then an awkward 3 or 4 seconds freeze. Though on subsequent launches only the scan was performed, so I think other actions came from version difference.

> With this PR, this empty scene is present but a progress bar is displayed in the FileSystem Dock. My concern is that the user can modify this empty scene and do a lot of things in the editor while the project file scan is in progress.

That's not really a problem, as the changes are not lost. If we really wanted to prevent that, the simplest solution is displaying a fully transparent Control over the whole editor that will block input.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -900,15 +924,23 @@ void EditorFileSystem::_scan_new_dir(EditorFileSystemDirectory *p_dir, Ref<DirAc
-				fi->type = ResourceFormatImporter::get_singleton()->get_resource_type(path);
-				fi->uid = ResourceFormatImporter::get_singleton()->get_resource_uid(path);
-				fi->import_group_file = ResourceFormatImporter::get_singleton()->get_import_group_file(path);
+				// Using get_resource_import_info to prevent calling 3 times ResourceFormatImporter::_get_path_and_type
```
```suggestion
				// Using get_resource_import_info() to prevent calling 3 times ResourceFormatImporter::_get_path_and_type.
```

Hilderin commented:
Thanks for the review. I definitively still need time to adjust to Godot standard. I'll try to be more careful next time.
All requested changes should be effective now.

Hilderin commented:
@KoBeWi thanks for testing this PR.

> I tested it with my project. I got FileSystemScan, then some new Scan Actions popup, then scene group scan popup and then an awkward 3 or 4 seconds freeze.

I'm intrigued with the _**3 or 4 seconds freeze**_, it's sounds like what @Zireael07 or @smix8 was explaining on the first load in #47053. Do you have the freeze after a reboot or the next day? I guess I can't have access to your project? Maybe you can explain to me briefly the amount and type of resources you have in your project so I can try to reproduce. Are you on Windows, Linux or Mac?

KoBeWi commented:
It might be related to plugins, autoloads and global scripts. Especially plugins, as I have many of them. Though it was a one-time thing, I'll see if it repeats tomorrow and report back.

KoBeWi commented:
I tested again and it still happened, but only once. Though it's not the same as reported by Zireael and smix, because it's much shorter. I think the original issue with unresponsive editor is covered by the new scan progresses, the remaining time can also come from restoring scenes.

I can send you the project if you want to test it, but it's quite big. DM me on the chat.

Hilderin commented:
After testing with the project from KoBeWi, I found 2 problems:

First: Update scene groups was triggered during reimportation because the ResourceImporterCSVTranslation is calling ResourceLoader.save() which trigger update_files. I added a condition to prevent updating script and scene groups during importation. It is done at the end of first scan.
That was causing a double progress:
![image](https://github.com/godotengine/godot/assets/81109165/6e7bc7e3-5a82-43f6-8306-e5c0f58334fc)
![image](https://github.com/godotengine/godot/assets/81109165/384b3bb3-bdfb-41ce-bf62-c583fb4506f5)

Second: As KoBeWi suspected, the froze editor was caused by the reopening of the scenes. On my PC, it takes 8 secs processing the first `sources_changed` caused by `EditorNode::_load_editor_layout`. I added another progress popup to prevent the frozen effect.
![image](https://github.com/godotengine/godot/assets/81109165/cab2700c-4663-4ad8-93df-8d45dac3048b)
![image](https://github.com/godotengine/godot/assets/81109165/bec75313-c882-4b14-9d41-ee3632e24815)


That is what is looks now:

https://github.com/godotengine/godot/assets/81109165/ea1ac2b2-22e1-469e-9a9a-0a543bbda229

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1743,9 +1794,13 @@ void EditorFileSystem::update_file(const String &p_file) {
+		if (!first_scan) {
+			_update_pending_script_classes();
+			_update_pending_scene_groups();
+			call_deferred(SNAME("emit_signal"), "filesystem_changed"); //update later
```
```suggestion
			call_deferred(SNAME("emit_signal"), "filesystem_changed"); // Update later.
```

Hilderin replied:
Done!

Hilderin commented:
Just rebased following conflicts.

akien-mga commented:
Thanks!

akien-mga commented:
Apologies, I jumped the gun on this PR which touches very core parts of the engine, prone to regressions and subtle bugs. I trusted existing reviews and forgot to assess how risky the changes are myself.

I think this should be kept for 4.4 as there will be likely be a number of follow-up issues that will be best handled in the 4.4 dev cycle than now as we try to stabilize 4.3.

So I'll revert the merge for 4.3, and would suggest reopening a PR with the same changes, that will target 4.4 and require more testing before merge.

Hilderin commented:
No problem @akien-mga , I'll reopen later!
